qmail-qfilter.c

```diff
@@ -0,0 +1,330 @@
+/* This is free and unencumbered software released into the public * domain. */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <signal.h>
+#include <unistd.h>
+
+#include "qmail-qfilter.h"
+
+#ifndef TMPDIR
+#define TMPDIR "/tmp"
+#endif
+
+#ifndef BUFSIZE
+#define BUFSIZE 4096
+#endif
+
+#define QQ_OOM 51
+#define QQ_WRITE_ERROR 53
+#define QQ_INTERNAL 81
+#define QQ_BAD_ENV 91
+
+#define QQ_DROP_MSG 99
+
+#define MSGIN 0
+#define MSGOUT 1
+#define ENVIN 3
+#define ENVOUT 4
+#define QQFD 5
+
+static const char* qqargv[2];
+
+/* a replacement for setenv(3) for systems that don't have one */
+void mysetenv(const char* key, const char* val, size_t vallen)
+{
+  char* tmp;
+  size_t keylen;
+  
+  keylen = strlen(key);
+  tmp = malloc(keylen + 1 + vallen + 1);
+  memcpy(tmp, key, keylen);
+  tmp[keylen] = '=';
+  memcpy(tmp+keylen+1, val, vallen);
+  tmp[keylen+1+vallen] = 0;
+  if (putenv(tmp) != 0)
+    exit(QQ_OOM);
+}
+
+void mysetenvu(const char* key, unsigned long val)
+{
+  char buf[40];
+  int i;
+  i = sizeof buf;
+  buf[--i] = 0;
+  do {
+    buf[--i] = (val % 10) + '0';
+    val /= 10;
+  } while (val > 0);
+  mysetenv(key, buf + i, sizeof buf - 1 - i);
+}
+
+static size_t env_len = 0;
+static size_t msg_len = 0;
+
+/* Parse the sender address into user and host portions */
+size_t parse_sender(const char* env)
+{
+  const char* ptr = env;
+  char* at;
+  size_t len = strlen(env);
+  
+  if(*ptr != 'F')
+    exit(QQ_BAD_ENV);
+  ++ptr;
+  
+  unsetenv("QMAILNAME");
+  
+  if(!*ptr) {
+    if (putenv("QMAILUSER=") != 0
+        || putenv("QMAILHOST=") != 0)
+      exit(QQ_OOM);
+    return 2;
+  }
+
+  at = strrchr(ptr, '@');
+  if(!at) {
+    len = strlen(ptr);
+    mysetenv("QMAILUSER", ptr, len);
+    putenv("QMAILHOST=");
+  }
+  else {
+    len = strlen(at);
+    mysetenv("QMAILUSER", ptr, at-ptr);
+    mysetenv("QMAILHOST", at+1, len-1);
+    ptr = at;
+  }
+  return ptr + len + 1 - env;
+}
+
+void parse_rcpts(const char* env, int offset)
+{
+  size_t len = env_len - offset;
+  const char* ptr = env + offset;
+  char* buf = malloc(len);
+  char* tmp = buf;
+  unsigned long count;
+  count = 0;
+  while(ptr < env + env_len && *ptr == 'T') {
+    size_t rcptlen = strlen(++ptr);
+    memcpy(tmp, ptr, rcptlen);
+    tmp[rcptlen] = '\n';
+    tmp += rcptlen + 1;
+    ptr += rcptlen + 1;
+    ++count;
+  }
+  *tmp = 0;
+  mysetenv("QMAILRCPTS", buf, tmp-buf);
+  mysetenvu("NUMRCPTS", count);
+  free(buf);
+}
+
+void parse_envelope(void)
+{
+  const char* env;
+  size_t offset;
+  if ((env = mmap(0, env_len, PROT_READ, MAP_PRIVATE, ENVIN, 0)) == MAP_FAILED)
+    exit(QQ_OOM);
+  offset = parse_sender(env);
+  parse_rcpts(env, offset);
+  munmap((void*)env, env_len);
+}
+
+/* Create a temporary invisible file opened for read/write */
+int mktmpfile()
+{
+  char filename[sizeof(TMPDIR)+19] = TMPDIR "/fixheaders.XXXXXX";
```

Comment:
The size of the buffer is not needed, when it is initialized it will automatically get the correct size. This will as well remove the magic 19.